lib/Adaptor.js

```diff
@@ -428,3 +431,139 @@ function upsertMany(table, uuid, records) {
+      return new Promise(function (resolve, reject) {
+        console.log("Describing table via : ".concat(query));
+        client.query(query, function (err, result) {
+          if (err) {
+            reject(err);
+            client.end();
+          } else {
+            //console.log(result);
+            resolve(result);
+          }
+        });
+      }).then(function (data) {
+        return _objectSpread({}, state, {
+          table_data: {
+            body: data
+          }
+        });
+      });
+    } catch (e) {
+      console.log(e);
+      client.end();
+    }
+  };
+}
+/**
+ * Create a table in database when given a form definition and a table_name.
+ * @example
+ * insertTable(
+ *  state => state.data.koboColumns, 'table_name'
+ * )
+ * @constructor
+ * @param {function} records - An array of form columns
+ * @param {function} table - The new table to create
+ * @returns {Operation}
+ */
+
+
+function insertTable(records, table) {
+  return function (state) {
+    var client = state.client;
+
+    try {
+      var recordData = records(state);
+      var structureData = recordData.map(function (x) {
+        return "".concat(x.name, " ").concat(x.type, " ").concat(x.required ? 'NOT NULL' : '');
+      }).join(', ');
+      var query = "CREATE TABLE ".concat(table, " (\n        ").concat(structureData, "\n      );");
+      return new Promise(function (resolve, reject) {
+        console.log("Creating table via : ".concat(query));
+        client.query(query, function (err, result) {
+          if (err) {
+            reject(err);
+            client.end();
+          } else {
+            console.log(result);
+            resolve(result);
+          }
+        });
+      }).then(function (data) {
+        return _objectSpread({}, state, {
+          response: {
+            body: data
+          }
+        });
+      });
+    } catch (e) {
+      console.log(e);
+      client.end();
+    }
+  };
+}
+/**
+ * Alter an existing table in the database.
+ * @example
+ * modifyTable(
+ *  state => state.data.koboColumns, 'table_name'
+ * )
+ * @constructor
+ * @param {function} records - An array of form columns
+ * @param {string} table - The name of the table to alter
+ * @returns {Operation}
+ */
+
+
+function modifyTable(records, table) {
+  return function (state) {
+    var client = state.client;
+
+    try {
+      var recordData = records(state);
+      var structureData = recordData.map(function (x) {
+        return "ADD COLUMN ".concat(x.name, " ").concat(x.type, " ").concat(x.required ? 'NOT NULL' : '');
+      }).join(', ');
+      var query = "ALTER TABLE ".concat(table, "\n        ").concat(structureData, "\n      ;");
+      return new Promise(function (resolve, reject) {
+        console.log("Altering table via : ".concat(query));
+        client.query(query, function (err, result) {
+          if (err) {
+            reject(err);
```

Comment:
I'm confused by this pattern in all three functions. Have you confirmed that it works for postgres server-side errors? Seeing the following:
```js
try {
  ...
  reject(err);
  client.end();
   ...
} catch (e) {
  ...
  client.end;
}

```
I would almost think that (a) these adaptors show :heavy_check_mark: _even_ when they should fail and (b) that when a server-side error occurs, we attempt to `end()` the `client` _twice_, rather than just once and that _this_ might throw another error.

Reply:
Hmm, here I tried to keep it uniform accross the whole adaptor. That's why I used this formatting. Otherwise we might to refactor all the helpers in there. What do you think?

Reply:
Let's discuss quickly after the standup today. Have you confirmed that this strategy works to produce the right exit codes under the right error/success conditions?